### PR TITLE
Cluster Controller updates

### PR DESCRIPTION
**Pull request by ckreibich**
This expands the cluster controller feature set over what we released with 4.1. It's not yet the full functionality I'm hoping to have in 4.2. It includes:

- Support for the new external cluster testsuite in a Github workflow. This runs as an additional job after our Docker image build, reusing that image for docker-compose setups. The external testsuite is at https://github.com/zeek/zeek-testing-cluster. This integration is missing bits still since it's not yet pinning the external testsuite to a particular version. We can use a mechanism similar to the existing `commit-hash` files (with one difference being that their use is currently tied to Cirrus, but we run the above in Github due to its much easier use of `docker-compose`).
- Consistent use of `ClusterController::Types::Result` records for conveying request results
- A timeout mechanism for request state, via a new `ClusterController::Request::request_expired` event
- A rework of the logic that controls communication between controller and agents/instances: instead of requiring explicit `redef`-level configuration of the controller/agent connectivity, this now happens implicitly via cluster layouts pushed via the client. This also provides an explicit `ClusterController::API::notify_agents_ready` event — this means readiness of the controller/agent connectivity, not yet that of the full data cluster.
- Minor tweaks.

Apologies for the size of this — I'll strive to keep the changesets smaller from now on. It'll get easier since things have fallen into place a bit.

zeek/zeek-client#1 is the companion PR to this one, providing the corresponding `zeek-client` updates.

### CODE REVIEW

File: scripts/policy/frameworks/cluster/agent/api.zeek [hunk 1]
```diff
@@ -30,4 +30,11 @@ export {
 
 	# Report informational message.
 	global notify_log: event(instance: string, msg: string, node: string &default="");
+
+	# Notification events, controller -> agent
+
+	# Confirmation from controller in response to notify_agent_hello
+	# that the agent is welcome.
+	global notify_controller_hello: event(controller: string, host: addr);
```
[the remark this answers is not in the document]
Comment (0xxon) [reply scripts/policy/frameworks/cluster/agent/api.zeek:38]: The same applies to the other events in this - and the other globals. Even though they were not changed in this PR, it would be neat to have them documented.

File: scripts/policy/frameworks/cluster/controller/api.zeek [hunk 2]
```diff
@@ -5,12 +5,27 @@ module ClusterController::API;
 export {
 	const version = 1;
 
+	# Triggered when the operational instances align with desired ones, as
+        # defined by the latest cluster config sent by the client.
```
Comment (0xxon) [inline scripts/policy/frameworks/cluster/controller/api.zeek:9]: indentation :)

File: scripts/policy/frameworks/cluster/controller/api.zeek [hunk 3]
```diff
@@ -5,12 +5,27 @@ module ClusterController::API;
...
 	    config: ClusterController::Types::Configuration);
 	global set_configuration_response: event(reqid: string,
 	    result: ClusterController::Types::ResultVec);
-}
+
+	# Testing events. These don't provide operational value but expose
+	# internal functionality, triggered by test cases.
+
+	# This event causes no further action (other than getting logged) if
+        # set_state is F. When T, the controller establishes request state. The
```
Comment (0xxon) [inline scripts/policy/frameworks/cluster/controller/api.zeek:25]: indentation is off starting here.

File: scripts/policy/frameworks/cluster/controller/main.zeek [hunk 4]
```diff
@@ -132,7 +265,13 @@ event ClusterAgent::API::set_configuration_response(reqid: string, result: Clust
 		ClusterController::Request::finish(r$id);
 		}
 
-	ClusterController::Log::info(fmt("tx ClusterController::API::set_configuration_response %s", req$id));
+	# This is the point where we're really done with the original
+        # set_configuration request. We adopt the configuration as the current
```
Comment (0xxon) [inline scripts/policy/frameworks/cluster/controller/main.zeek:269]: indentation :)

File: scripts/policy/frameworks/cluster/controller/util.zeek [hunk 5]
```diff
@@ -0,0 +1,17 @@
+module ClusterController::Util;
+
+export {
+	global set_to_vector: function(ss: set[string]): vector of string;
```
Comment (0xxon) [inline scripts/policy/frameworks/cluster/controller/util.zeek:4]: This should be documented. Also - should this one perhaps be renamed something like string_set_to_vector and moved to `base/utils/strings.zeek`?

**Comment by timwoj** [comment]
> I admittedly mostly concentrated on stylistic issues

As a note, we do have python formatting via `pre-commit` now. You might need to rebase the branch to pull in the support for it, if it's been a while since you last did that.

File: scripts/policy/frameworks/cluster/controller/config.zeek [hunk 6]
```diff
@@ -28,17 +28,15 @@ export {
...
-	const instances: table[string] of ClusterController::Types::Instance = { } &redef;
-
 	# The role of this node in cluster management. Agent and
 	# controller both redef this. Used during logging.
 	const role = ClusterController::Types::NONE &redef;
 
+	# The timeout for request state. Applies both to state kept for client
+	# requests, as well as state in the agents for requests to the
+	# supervisor.
+	const request_timeout = 10sec &redef;
```
Comment (0xxon) [inline scripts/policy/frameworks/cluster/controller/config.zeek:38]: This is also not in a doxygen-parseable format.

Generally, I think all entries in this file should be changed to be recognized by doxygen - otherwise the documentation for this feels very barren. I think we should do that either in this PR, or before the release is cut.

File: .github/workflows/docker.yml [hunk 7]
```diff
@@ -13,8 +13,12 @@ defaults:
   run:
     shell: bash
 
+env:
```
Comment (timwoj) [inline .github/workflows/docker.yml:16]: Is this whole commit about docker related to this PR or can it be merged separately?

**Review by timwoj — COMMENTED** [review]
I tried to follow the control a bit during my review. It all seems reasonable to me. I don't have any additional comments other than piggy-backing on a couple of @0xxon's.

**Comment by ckreibich** [comment]
Thanks for jumping on this, folks. I'll sort out the indentation — my edit was masking it, ewps.

I hadn't considered the structure sufficiently settled for spending much time on zeekygen-level commenting, but maybe I can't make that excuse any longer. :)

And, I'll look into the CI failures. I'm thrilled though that the `docker-compose` one is running!

**Comment by ckreibich** [comment]
I've resolved the indentation and changed the naming of globals. I'm working on a documentation pass.

File: scripts/policy/frameworks/cluster/controller/main.zeek [hunk 8]
```diff
@@ -333,51 +376,37 @@ event ClusterController::API::get_instances_request(reqid: string)
...
 	Broker::subscribe(ClusterController::topic);
 
+	# Events sent to the client:
+
 	Broker::auto_publish(ClusterController::topic,
 	    ClusterController::API::get_instances_response);
 	Broker::auto_publish(ClusterController::topic,
 	    ClusterController::API::set_configuration_response);
 
-	if ( |ClusterController::instances| > 0 )
```
Comment (rsmmr) [inline scripts/policy/frameworks/cluster/controller/main.zeek:226]: There's no initial peering anymore at controller startup now. Is that because we won't have any instances at first now? Once we get config persistence, we'll get that back I suppose?

File: scripts/policy/frameworks/cluster/agent/main.zeek [hunk 9]
```diff
@@ -22,6 +22,12 @@ global g_nodes: table[string] of ClusterController::Types::Node;
 # new configurations.
 global g_data_cluster: table[string] of Supervisor::ClusterEndpoint;
 
+# Whether we currenty keep notifying the controller that we're here.
+# We stop once the controller responds back, and resume when we've
+# lost the peering.
```
Comment (rsmmr) [inline scripts/policy/frameworks/cluster/agent/main.zeek:29]: I can't quite align the comment with the code: "keep notifying" suggest to me that we continue to send the even unsolicited, but if I'm reading the code right, we send it only on demand after receiving the hello from cluster.

File: scripts/policy/frameworks/cluster/agent/main.zeek [hunk 10]
```diff
@@ -161,14 +167,28 @@ event ClusterAgent::API::set_configuration_request(reqid: string, config: Cluste
...
+event ClusterAgent::API::notify_controller_hello(controller: string, host: addr)
+	{
+	ClusterController::Log::info(fmt("rx ClusterAgent::API::notify_controller_hello %s %s", controller, host));
+	g_notify_controller = F;
+	}
+
+event ClusterAgent::API::notify_agent_hello(instance: string, host: addr, api_version: count)
+	{
+	if ( g_notify_controller )
+		schedule 1sec { ClusterAgent::API::notify_agent_hello(instance, host, api_version) };
```
Comment (rsmmr) [inline scripts/policy/frameworks/cluster/agent/main.zeek:181]: Just curious: why a schedule of 1sec vs immediately?

File: scripts/policy/frameworks/cluster/controller/main.zeek [hunk 11]
```diff
@@ -230,91 +279,78 @@ event ClusterController::API::set_configuration_request(reqid: string, config: C
...
+		if ( inst$name !in insts_to_keep )
+			next;
+
+		# XXX 0.0.0.0 is a workaround until we've resolved
+		# how agents that peer with us obtain their identity.
+		# Broker ID?
+		if ( ( inst$host != 0.0.0.0 && inst$host != g_instances[inst$name]$host ) ||
+		     inst?$listen_port != g_instances[inst$name]?$listen_port ||
+		     ( inst?$listen_port && g_instances[inst$name]?$listen_port &&
+		       inst$listen_port != g_instances[inst$name]$listen_port ) )
```
Comment (rsmmr) [inline scripts/policy/frameworks/cluster/controller/main.zeek:363]: Worth factoring this out into a comparison function for clarity?

**Review by rsmmr — COMMENTED** [review]
Nice how this is shaping up. In some sense I think this is the most important logic of all: the basic handling of connectivity & configuration between the various players. Getting this ready and robust, will be a great foundation for everything else (my comment on double-checking paths is coming from that perspective).

I didn't look much further at the testing setup, but I like what you had shown/explained to me before.

Couple more things:

- Agree with Johanna's note on doc strings, and also agree that this doesn't need to block the PR; can do before final release.

- One challenge I keep noticing reading over the code is keeping it straight in my head where events get raised vs processed. Specifically. I keep looking at event handler and wondering (1) is the coming from the local process or a remote process, and (2) is this actively triggering some action here for this process, or is just something like logging and the action is somewhere else. What would help I think is if we could up with some convention to express origin and action semantics through either the event names (we are encoding the request/response in there already, but that isn't quite doing it for me yet) or by just putting a comment in front of the handler summarizing briefly.

**Comment by rsmmr** [comment]
Adddemdum: I looked only over the commits yet. I'm still want to play with it a bit too, hopefully later today, or otherwise tomorrow.

**Comment by rsmmr** [comment]
So just a couple more things I noticed when running the client:

- If I have no controller running, simply executing `zeek-client` (no arguments) hangs for me completely, I can't even CTRL-C it. Tracing the script, it goes into the `peer()` call right away. If I have a controller running, I get an error message instead (`error: please provide a command to execute. See --help.`). So I guess it would be good to start peering only once needed (the lack of interuptability is presumably a Broker issue).
- After a successful `set-config`, the client tells me `WARNING  received unexpected event "ClusterController::API::notify_agents_ready", skipping`

**Comment by ckreibich** [comment]
This is really great, thank you for digging in Robin!

> One challenge I keep noticing reading over the code is keeping it straight

:-) — yes, completely agree. The eventing makes the flow hard to follow. I suspect that's also what Johanna was referring to. For example, when you push a config an event comes in from the client, needs to be relayed to the agents after some examination, and once they all have responded (or things time out), you respond back to the client. But the reader has to figure that out ... At one point I wondered whether it'd be worth writing a lightweight Zeek meta DSL to express this, but I it'd be a big detour. For now it'll certainly help to explain the flow in documentation. I also think it'll be easier to see the patterns once they happen for more client commands.

In terms of naming, currently the only conventions are only these:

- if it's in the `ClusterAgent::API` space it's controller/agent communication, if it's in `ClusterController::API` it's controller/client
- request/response pairs are coupled accordingly in their naming.  

I'll follow up on your specific comments next.

File: scripts/policy/frameworks/cluster/controller/main.zeek [hunk 12]
```diff
@@ -9,20 +9,46 @@
 
 redef ClusterController::role = ClusterController::Types::CONTROLLER;
 
+global config_current: ClusterController::Types::Configuration;
+global config_reqid_pending: string = "";
+
+function send_config_to_agents(req: ClusterController::Request::Request,
```
[the remark this answers is not in the document]
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/controller/main.zeek:42]: Agreed, it's very possible there are gaps. I've found this tricky to test — for example, for testing you'd like to say "okay agent, please now don't actually respond to this", or alternatively "okay let's ignore this response" now. I've been reluctant to build such testability directly into the code. It's good fodder for our unit-test discussion — what would a harness look like that can handle the controller in isolation and just feed/receive events? Etc.

File: scripts/policy/frameworks/cluster/controller/main.zeek [hunk 13]
```diff
@@ -141,57 +199,89 @@ event ClusterController::API::set_configuration_request(reqid: string, config: C
...
 
+	local res: ClusterController::Types::Result;
 	local req = ClusterController::Request::create(reqid);
-	req$set_configuration_state = ClusterController::Request::SetConfigurationState();
+	req$set_configuration_state = ClusterController::Request::SetConfigurationState($config = config);
 
-	# Compare new configuration to the current one and send updates
-	# to the instances as needed.
-	if ( config?$instances )
+	# At the moment there can only be one pending request.
```
[the remark this answers is not in the document]
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/controller/main.zeek:290]: It'd certainly be better. Right now the idea is simply that if you submit a request while the controller hasn't fully processed the former, you're simply told that fact and rejected. It's another foundation building block because I'm sure such request queueing will come in handy for other types of requests.

File: scripts/policy/frameworks/cluster/controller/main.zeek [hunk 14]
```diff
@@ -141,57 +199,89 @@ event ClusterController::API::set_configuration_request(reqid: string, config: C
...
+			if ( ! inst?$listen_port )
 				{
-				local res = ClusterController::Types::Result($reqid=reqid, $instance=inst$name);
-				res$error = fmt("instance %s is unknown, skipping", inst$name);
+				# If a requested instance doesn't have a listen port and isn't known
+				# to us, we have no way to establish connectivity. We reject the
+				# request.
+				res = ClusterController::Types::Result($reqid=reqid, $instance=inst$name);
+				res$success = F;
+				res$error = fmt("instance %s is unknown", inst$name);
```
[the remark this answers is not in the document]
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/controller/main.zeek]: This is from an earlier version of the code, right? In the current code, that logic has changed with the move to defining the agents/instances via the pushed confuration. The relevant code is now in the `notify_agent_hello` event handler, which [logs the fact](https://github.com/zeek/zeek/blob/topic/christian/cluster-controller-next/scripts/policy/frameworks/cluster/controller/main.zeek#L160-L162). (Happy to change the wording of course!)

File: scripts/policy/frameworks/cluster/agent/main.zeek [hunk 15]
```diff
@@ -9,18 +9,18 @@
 redef ClusterController::role = ClusterController::Types::AGENT;
 
 # The global configuration as passed to us by the controller
-global global_config: ClusterController::Types::Configuration;
+global g_config: ClusterController::Types::Configuration;
```
[the remark this answers is not in the document]
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/agent/main.zeek:14]: Yeah that's nice. Atm there are globals for the agent's runtime (`main.zeek`), the controller's, and for the request abstraction in the `ClusterController::Request` module. I'll keep an eye on it.

Btw, I at times wondered whether it'd be good to move code that both agent and controller rely on to a separate location. I was torn about naming so didn't do it. For example, requests live in `ClusterController::Requests`, but it's also used by the agent to track queries out to the supervisor and back. It might be worth pondering a namespace layout that accommodates the management cluster (what we're building here) and the data cluster ("the cluster" so far, `Cluster::`), to make this more natural. Just to throw something out there: we could use `Management::Controller` and `Management::Agent`, and then have e.g. `Management::Requests` not tied to controller or agent. I also like "management framework" as a name for this whole thing.

File: scripts/policy/frameworks/cluster/controller/main.zeek (hunk 8 above)
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/controller/main.zeek:226]: Yes exactly — with persistence, there'll be something to do, right now it all waits until you push a config.

File: scripts/policy/frameworks/cluster/agent/main.zeek (hunk 9 above)
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/agent/main.zeek:29]: It should work as follows:
- Regardless of the direction of connection establishment, the `Broker::peer_added` event fires upon connectivity, and the agent starts sending the `notify_agent_hello` events. (There's no point in sending them before there's connectivity.)
- These pings continue until the controller sends `notify_controller_hello`, which is the signal to the agent that it's now welcome, as per the currently active configuration. The agent stops pinging, and will resume upon the next `peer_added`.

But! Looking at this now, seems we can simplify: the point of `notify_agent_hello` with the current approach is really to convey to the controller that an agent is ready. There's no need for this ongoing pinging: we could have a request/response event pair `agent_welcome_request` and `agent_welcome_response` ... the controller sends the request to agents it wants to see in service, and tracks the response event like it now tracks `notify_agent_hello`. And for agents the controller no longer requires it can send a `agent_farewell`, no response needed, which'll cause the agent to stop its cluster nodes.

What do you think? I'll give this a go now.

File: scripts/policy/frameworks/cluster/agent/main.zeek (hunk 10 above)
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/agent/main.zeek:181]: It was to build ongoing pings to the controller at a reasonable rate ... but per the above, I think we should rework this.

File: scripts/policy/frameworks/cluster/controller/main.zeek (hunk 11 above)
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/controller/main.zeek:363]: Good one, will do. That one's unwieldy.

**Comment by ckreibich** [comment]
* If I have no controller running, simply executing zeek-client (no arguments) hangs for me completely

D'oh! I've never done this. Thank you! I'll look into it.

* After a successful set-config, the client tells me WARNING received unexpected event "ClusterController::API::notify_agents_ready", skipping

Ah yeah — this means that the client received an event it doesn't currently know what to do with. I'll add it. It's intentional that the controller sends it on, since it seems useful for the client to be able to tap into.

File: scripts/policy/frameworks/cluster/controller/main.zeek (hunk 14 above)
[the remark this answers is not in the document]
Comment (rsmmr) [reply scripts/policy/frameworks/cluster/controller/main.zeek]: Must have been one of the earlier commits, seems fine now.

File: scripts/policy/frameworks/cluster/controller/main.zeek (hunk 12 above)
[the remark this answers is not in the document]
Comment (rsmmr) [reply scripts/policy/frameworks/cluster/controller/main.zeek:42]: Yeah, definitely worth more thinking. For now, let me ask this way: what state is the code currently in? Have you already tried to cover all the error cases to the degree you could identify them (even wo/ specifically testing it out), or is it for now more "let's get it to work and do the robustness fine-tuning later"? if the latter, that's fine, then we don't need to do much more right now. If the former, I could do a pass next week for a 2nd set of eyes and see if I catch anything that skipped through (I'd need to follow the flow more closely than I have done for the review).

File: scripts/policy/frameworks/cluster/agent/main.zeek (hunk 15 above)
[the remark this answers is not in the document]
Comment (rsmmr) [reply scripts/policy/frameworks/cluster/agent/main.zeek:14]: > Btw, I at times wondered whether it'd be good to move code that both agent and controller rely on to a separate location. 

Good point. I had been wondering about that too, but couldn't quite pinpoint yet what to suggest; same for the naming on existing cluster framework vs the new controller layer.

> Just to throw something out there: we could use `Management::Controller` and `Management::Agent`, and then have e.g. `Management::Requests` not tied to controller or agent. I also like "management framework" as a name for this whole thing.

I like "management". Also, I've been starting to wonder if "ClusterController" is too confusing given the existing ZeekControl. So, yes, could work well to refer to the whole thing as "mgmt framework" and the controller being just an internal piece of it.

File: scripts/policy/frameworks/cluster/agent/main.zeek (hunk 9 above)
Comment (rsmmr) [reply scripts/policy/frameworks/cluster/agent/main.zeek:29]: Yep, that simplification sounds good (and the description makes sense, I must have gotten confused with event semantics, per the other comment :-)

**Comment by rsmmr** [comment]
Fine with me to merge this as is at this point, and then iterate. It's not affecting anything existing, so that shouldn't be a problem.

File: scripts/policy/frameworks/cluster/controller/main.zeek (hunk 12 above)
[the remark this answers is not in the document]
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/controller/main.zeek:55]: Sorry for the late reply here Robin — the answer is that I have tried to call our error cases as they came to mind via `XXX` comments in the code, and I've at this point taken care of some but not all. For example, you probably noticed that the controller's `set_configuration` implementation has such a comment for validation that it should do on the received configuration prior to going ahead with it. It would be very useful if you could note down possible corner cases, problem spots, targets of misuse, etc while you read this or tinker with it.

File: scripts/policy/frameworks/cluster/agent/main.zeek (hunk 9 above)
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/agent/main.zeek]: I've now worked this in. I simply put the change in an additional commit since I got afraid that it's too painful to sort through each subsequent commit if I rebase. `notify_agent_hello` is now really just a checkin to convey after a peering that we're dealing with an agent, and the controller uses `agent_welcome` request/responses to notify an agent that it's now going to participate in the running cluster, while a `agent_standby` request/response tells it to free it's cluster state and wait for further instructions.

There's one more minor change in this: I'm no longer trying to `unpeer()` unused agents, i.e., an agent can simply stick around even if it's not currently useful. This was imperfect anyway since one can only do this if one also previously `peer()`'ed, and it seems that when managing agents this way, there'll often be a need to restart it anyway.

I remembered that Benjamin has a nice way of visualizing protocol interactions via `plantuml` ... I'll see if I can get that to work so I can use it in comments.

File: scripts/policy/frameworks/cluster/controller/main.zeek (hunk 11 above)
Comment (ckreibich) [reply scripts/policy/frameworks/cluster/controller/main.zeek]: Done, this moved to the new `is_instance_connectivity_change()` function in the same file.

File: scripts/policy/frameworks/cluster/agent/main.zeek (hunk 9 above)
Comment (rsmmr) [reply scripts/policy/frameworks/cluster/agent/main.zeek]: Looks good!
